Analysis/Tasks/PWGHF/HFCandidateCreatorBPlus.cxx

```diff
@@ -0,0 +1,286 @@
+
+      //loop over tracks for pi selection
+      auto count = 0;
+      for (auto& track : tracks) {
+        //if(count % 100 == 0){
+        //LOGF(INFO, "Col: %d (cand); %d (track)", candidate.collisionId(), track.collisionId());
+        //  count++;
+        // }
+
+        if (cutEtaTrkMax >= 0. && std::abs(track.eta()) > cutEtaTrkMax) {
+          continue;
+        }
+
+        if (candidate.index0Id() == track.globalIndex() || candidate.index1Id() == track.globalIndex()) {
+          continue; //daughter track id and bachelor track id not the same
+        }
+
+        //Select D0pi- and D0(bar)pi+ pairs only
+        if (!((candidate.isSelD0() >= d_selectionFlagD0 && track.sign() < 0) || (candidate.isSelD0bar() >= d_selectionFlagD0bar && track.sign() > 0))) {
+          //Printf("D0: %d, D0bar%d, sign: %d", candidate.isSelD0(), candidate.isSelD0bar(), track.sign());
+          continue;
+        }
+
+        auto trackBach = getTrackParCov(track);
+        std::array<float, 3> pVecD0 = {0., 0., 0.};
+        std::array<float, 3> pVecBach = {0., 0., 0.};
+        std::array<float, 3> pVecBCand = {0., 0., 0.};
+
+        //find the DCA between the D0 and the bachelor track, for B+
+        if (bfitter.process(trackD0, trackBach) == 0) {
+          continue;
+        }
+
+        bfitter.propagateTracksToVertex();          // propagate the bachelor and D0 to the B+ vertex
+        bfitter.getTrack(0).getPxPyPzGlo(pVecD0);   //momentum of D0 at the B+ vertex
+        bfitter.getTrack(1).getPxPyPzGlo(pVecBach); //momentum of pi+ at the B+ vertex
+        const auto& BSecVertex = bfitter.getPCACandidate();
+        auto chi2PCA = bfitter.getChi2AtPCACandidate();
+        auto covMatrixPCA = bfitter.calcPCACovMatrix().Array();
+        hCovSVXX->Fill(covMatrixPCA[0]); //FIXME: Calculation of errorDecayLength(XY) gives wrong values without this line.
+
+        pVecBCand = RecoDecay::PVec(pVecD0, pVecBach);
+
+        // get track impact parameters
+        // This modifies track momenta!
+        auto primaryVertex = getPrimaryVertex(collision);
+        auto covMatrixPV = primaryVertex.getCov();
+        hCovPVXX->Fill(covMatrixPV[0]);
+        o2::dataformats::DCA impactParameter0;
+        o2::dataformats::DCA impactParameter1;
+
+        bfitter.getTrack(0).propagateToDCA(primaryVertex, d_bz, &impactParameter0);
+        bfitter.getTrack(1).propagateToDCA(primaryVertex, d_bz, &impactParameter1);
+
+        // get uncertainty of the decay length
+        double phi, theta;
+        getPointDirection(array{collision.posX(), collision.posY(), collision.posZ()}, BSecVertex, phi, theta);
+        auto errorDecayLength = std::sqrt(getRotatedCovMatrixXX(covMatrixPV, phi, theta) + getRotatedCovMatrixXX(covMatrixPCA, phi, theta));
+        auto errorDecayLengthXY = std::sqrt(getRotatedCovMatrixXX(covMatrixPV, phi, 0.) + getRotatedCovMatrixXX(covMatrixPCA, phi, 0.));
+
+        int hfFlag = 1 << hf_cand_bplus::DecayType::BPlusToD0Pi;
```

Comment:
Yes it is used in HFBPlusToD0PiCandidateSelector.cxx and taskBPlus.cxx, similar to D0 task, to select B+ candidates decaying to D0pi. 

For the previous comment, I cant add a reply, so I add here - "The "df" object is used to reconstruct D0 secondary vertex and get it CovMatrix. The "bfitter“ object is used for B+ vertex reconstruction, propagate daughters to the B vertex, and get its CovMatrix and other B properties.  Since df corresponds to D0 vertex and bfitter corresponds to B+ vertex, I am not sure how I can use the same object for both."